### ISSUE TITLE
Edgedb schema import/export

### ISSUE THREAD

**Issue by MrNaif2018** [edgedb/edgedb#702]
Is it possible to load(import migrations from file) or to save(export) migrations to .esdl file?
As far as I seen, not yet.
I want to implement this feature, but I need to discuss how would it work first.
My idea is to integrate it to cli like so:
`edgedb schema export file_name.esdl`
to export current database schema to file
`edgedb schema import file_name.esdl` to import to database
Where importing would just run
`CREATE MIGRATION {file_name} TO {file_contents}`
And exporting would save all the schema? From what I can see, there is only CREATE MIGRATION and GET MIGRATION, is it possible to get current schema or list of migrations via edgeql?

**Comment by elprans** [comment]
Hey @MrNaif2018!  This feature is on the [roadmap for 1.0](https://edgedb.com/roadmap/#sdl_and_migrations).  We are currently doing some preparatory work and refactoring needed to unblock this, including the ability to dump schema modules to `.esdl` files.

Roughly, the idea is that you *start* with `.esdl` files in your repo and use `edgedb migration` to generate and apply schema migrations.  This workflow is similar to that of Django and other such frameworks.  There will also be a certain level of interactivity for situations when the tool is not 100% sure about the schema change, e.g. to resolve renames vs drop/create operations etc.

We are also planning to reimplement the CLI tools in Golang for easier distribution, so we're holding off with major work on the current CLI frontend.

**Comment by MrNaif2018** [comment]
This approach similar to django is actually the only thing that stops me from using edgedb even now. Can I help somehow to speed this up by contributing?

**Comment by elprans** [comment]
Absolutely! Help is welcome. I'll open a new issue describing the overall design of `edgedb migration` and we can identify specific tasks there.

**Comment by MrNaif2018** [comment]
Will this issue be opened only after some refactor? I can start working on implementing this in any time.

**Comment by elprans** [comment]
Yes, I'll get to it once #705 and #710 are landed.

**Comment by agritheory** [comment]
@elprans I'd like to help with the python bindings for this @MrNaif2018 I'd be happy to coordinate efforts (as opposed to working alone)

**Comment by MrNaif2018** [comment]
I am happy to coordinate efforts too. Together we can create faster and better!

**Comment by agritheory** [comment]
I'm thinking that some portion of the parser ([maybe this part](https://github.com/edgedb/edgedb/blob/master/edb/edgeql/parser/__init__.py#L62)?) would need to move to the python client for import validation.

**Comment by elprans** [comment]
We definitely don't want to be copying parsers to the client.  Most functionality in the language drivers should be thin wrappers around backend API.

**Comment by agritheory** [comment]
Gotcha. I think I would need to wait for the top level API to be exposed before wrapping it on the client side then?

**Comment by MrNaif2018** [comment]
Also is there something I can help with as #710 issue which got closed? Maybe something from task list? (Cleanup issues?)

**Comment by elprans** [comment]
> Also is there something I can help with

Writing more migration tests (see #710 and #715 for examples) would be excellent help.

**Comment by elprans** [comment]
> I think I would need to wait for the top level API to be exposed before wrapping it on the client side then?

The `CREATE MIGRATION` command can already be used for the purpose of ESDL validation.

**Comment by agritheory** [comment]
@elprans Is there a spec for .edsl files? Expected encoding, new line conventions, new line at end of file... that kind of thing?

**Comment by elprans** [comment]
@agritheory SDL is actually lexically identical to EdgeQL.   SDL declarations are identical in syntax to equivalent DDL commands without the `CREATE` or `SET` keywords.  The lexical structure of EdgeQL/SDL is documented here: https://edgedb.com/docs/edgeql/lexical

**Comment by agritheory** [comment]
@elprans Thank you for that clarification, I'm getting more and more familiar with the edgeDB documentation every day. The intent of my original question was to ask if there were any specific encoding requirements for the file itself, like UTF-8/16 that need to be checked, and other things to make sure the file is syntactically correct. If this isn't spelled out anywhere, I can guess on some sane defaults and then they can be reviewed in a PR.

**Comment by elprans** [comment]
EdgeDB itself does not interact with *.esdl files directly, so we don't impose any restrictions how the file is actually encoded.  That said, I cannot think of any reason why anybody would want something other than UTF-8, as even Windows Notepad supports it nowadays.  As for the newlines, as long as it's a usual combination of `\n` and/or `\r`, it's fine.

**Comment by elprans** [comment]
To export current schema into `.esdl` files a `DESCRIBE MODULE blah AS SDL` command can be used.  The implementation of [RFC 1000](https://github.com/edgedb/rfcs/blob/master/text/1000-migrations.rst) in [edgedb/edgedb-cli](https://github.com/edgedb/edgedb-cli) would take care of the inverse (import/migrations).
